tools/jdbc/src/main/java/org/duckdb/DuckDBConnection.java

```diff
@@ -25,6 +25,8 @@
 public class DuckDBConnection implements java.sql.Connection {
 	protected ByteBuffer conn_ref = null;
 	protected DuckDBDatabase db;
+	protected boolean autoCommit = true;
```

Comment:
How does keeping this state in the JDBC client interface with users running `BEGIN TRANSACTION`, `COMMIT`, etc as regular queries instead of using the `.commit()` method? How should it behave?

Reply:
For the how it should behave I have two quotes:

> When to start a new transaction is a decision made implicitly by either the JDBC
> driver or the underlying data source. Although some data sources implement an
> explicit “begin transaction” statement, there is no JDBC API to do so. Typically, a
> new transaction is started when the current SQL statement requires one and there is
> no transaction already in place. 
[1]

> When auto-commit is disabled, each transaction must be explicitly committed by
> calling the Connection method commit or explicitly rolled back by calling the
> Connection method rollback, respectively.
[2]

From my point of view this is covered by the PR.

The question is what should happen if someone does not follow the specs. Two begin transactions throw an error. Same with commit/rollback. So only with auto-commit true it would be possible to explicitly start a transactions with a "BEGIN TRANSACTION" statement. For commit/rollback both ways would be possible. I would consider this as a good way to use one-off longer transaction without changing the auto-commit behavior, but it is necessary to check the auto-commit state before to avoid exceptions.

[1] JDBC 4.2 Specification, Chapter 10.1
[2] JDBC 4.2 Specification, Chapter 10.1.1

Reply:
Could we add tests that check this behavior, and ideally have this work as expected? You can check whether there is currently an active transaction by calling the `context.transaction.ActiveTransaction()` function. If this returns true, auto-commit is disabled (i.e. `BEGIN TRANSACTION` has been explicitly called). If this returns false, auto-commit is enabled.

Reply:
No problem, I can add some tests, but I am not exactly sure what behavior is expected:

If the "JDBC way" is used, everything works as expected &#8594; no problems here

Problems start with explicit `BEGIN TRANSACTION` If auto-commit is off, there will be an exception. This is correct as nesting is not allowed and this might be the user's intention. If auto-commit is on, the transaction will stay active which I believe is also correct.

Explicit `COMMIT` or `ROLLBACK` throw an exception if no transaction is active. This is independent of auto-commit and consistent with e.g. the CLI behavior.

We could switch auto-commit to off when we detect an active transaction. I am not sure if this is a good decision as it should be set via the methods and is a Connection property that outlives the transaction. The following transaction should not notice a difference in my eyes.

So my suggestion would be: For those who want to ignore the standard JDBC behavior, we can add a method to test for active transactions via the `context.transaction.ActiveTransaction()` and give them all means to do their own transaction handling.